### PR TITLE
Add del's to reductions.

### LINKED ISSUE #5003: Memory leak when allocating arrays in function in parallel loop
I tired to accelerate a numerical simulation using the excellent `nb.prange` for simple parallelization. At its core, my simulation calls a single function (`update` in the code below) many times during each timestep. This function updates the central data array in arbitrary (i.e. hard to predict) locations. I wanted to accelerate the inner loop using `nb.prange` and used a chunked iteration with temporary arrays to avoid the race condition when multiple threads write to the same data array. However, using the simplest implementation, I ran into a memory leak, which I consider a bug.

To illustrate the problem, I attach a working example, where calling `test(parallel)` shows the memory leak. The allocated memory seems to scale linearly with the number `rep` of repetitions. I saw this problem with the following Numba environment on MacOs 10.14.6 using python 3.7:

>version: 0.46.0
>using_svml: False
>threading_layer: tbb
>omp_num_threads: None
>mkl_num_threads: None
>num_threads: 12

If you cannot see the problem on your setup, it might help to increase `rep`, which makes the memory leak for longer on my setup. 

```python
#!/usr/bin/env python3

import numpy as np
import numba as nb


num = 10_000_000
rep = 1000
threads = 5
dim = 10_000


@nb.jit
def update(arr, out, offset=0):
    """ inner update function """
    for i in range(arr.size):
        out[(i + offset) % dim] += arr[i]


@nb.jit
def basic(arr, offset=0):
    """ basic usage of the update function """
    out = np.zeros(dim, np.int64)
    update(arr, out, offset)
    return out
     

@nb.jit(parallel=True)
def race_condition(arr):
    """ naive parallization with race condition """
    out = np.zeros(dim, np.int64)
    size = int(np.ceil(arr.size / threads))
    for i in nb.prange(threads):
        arr_junk = arr[i * size:(i + 1) * size]
        update(arr_junk, out, i * size)
    return out
     

def iter_in_junks(arr):
    """ junked iteration guarding against race condition """
    # get size of a single junk
    res = np.zeros(dim, np.int64)
    size = int(np.ceil(arr.size / threads))
    for i in nb.prange(threads):
        arr_junk = arr[i * size:(i + 1) * size]
        res += basic(arr_junk, i * size)
        
    return res
        
# compile without and with parallel support
single = nb.jit()(iter_in_junks)
parallel = nb.jit(parallel=True)(iter_in_junks)
        
arr = np.arange(num)
expect = basic(arr)
        
def test(func):
    """ test function that repeatably calls func and test result """
    
    @nb.jit
    def inner():
        for i in range(rep):
            res = func(arr)
            if not np.array_equal(res, expect):
                return False
        return True
        
    return inner()
        
# test the different variants
print('Race condition:', test(race_condition))
print('Single thread:', test(single))
print('Multiple threads:', test(parallel))
```

### PR DESCRIPTION
**Pull request by DrTodd13**
Resolves #5003

### CODE REVIEW

File: numba/npyufunc/parfor.py [hunk 1]
```diff
@@ -1539,4 +1552,7 @@ def load_range(v):
         only_elem_ptr = builder.gep(rv_arg, [context.get_constant(types.intp, 0)])
         builder.store(builder.load(only_elem_ptr), lowerer.getvar(k))
 
+#    for v in redarrdict.values():
+#        lowerer.lower_inst(ir.Del(v.name, loc=loc))
+
```
Comment (stuartarchibald) [inline numba/npyufunc/parfor.py]: dead code?

Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: This is the code that made me want to look at the PR again.  Want to make sure the leak appears to be gone.  I also had the same question about how to test this and I don't know the answer. @stuartarchibald

Comment (stuartarchibald) [reply numba/npyufunc/parfor.py]: ah ok, thanks. I'll put working out how to test this on my TODOs, I'm sure we can come up with something, even if we have to write something new it'd probably prove quite useful.

Comment (DrTodd13) [reply numba/npyufunc/parfor.py]: Removed those commented lines.

**Comment by DrTodd13** [comment]
@stuartarchibald What are you waiting for me on this PR?

**Comment by stuartarchibald** [comment]
> @stuartarchibald What are you waiting for me on this PR?

I think this is just waiting for a test, which either you are I need to work out how to do :) ?

**Pull request closed without merging**